v3/mathjax3-ts/core/InputJax.ts

```diff
@@ -0,0 +1,165 @@
+import {MmlNode} from './MmlTree/MmlNode.js';
+import {userOptions, defaultOptions, OptionList} from '../util/Options.js';
+import {FunctionList} from '../util/FunctionList.js';
+
+/*****************************************************************/
+/*
+ *  The InputJax interface
+ */
+
+export interface InputJax {
+    /*
+     * The name of the input jax subclass (e.g,. 'TeX')
+     */
+    name: string;
+
+    /*
+     * Whether this input jax processes string arrays or DOM nodes
+     * (TeX and AsciiMath process strings, MathML processes DOM nodes)
+     */
+    processStrings: boolean;
+
+    /*
+     * The options for this input jax instance
+     */
+    options: OptionList;
+
+    /*
+     * Lists of pre- and post-filters to call before and after processing the input
+     */
+    preFilters: FunctionList;
+    postFilters: FunctionList;
+
+    /*
+     * Finds the math within the DOM or the list of strings
+     *
+     * @param{Element | string[]} which  The element or array of strings to be searched for math
+     * @param{OptionList} options        The options for the search, if any
+     * @return{ProtoItem[]}              Array of proto math items found (further processed by the
+     *                                     handler to produce actual MathItem objects)
+     */
+    findMath(which: Element | string[], options?: OptionList): ProtoItem[];
+
+    /*
+     * Convert the math in a math item into the internal format
+     *
+     * @param{MathItem} math  The MathItem whose math content is to processed
+     * @return{MmlNode}       The resulting internal node tree for the math
+     */
+    compile(math: MathItem): MmlNode;
+}
+
+/*****************************************************************/
+/*
+ *  The InputJax class interface
+ */
+
+export interface InputJaxClass {
+    /*
+     *  The name of this subclass
+     */
+    NAME: string;
+
+    /*
+     *  The default options for this input jax class
+     */
+    OPTIONS: OptionList;
+
+    new(options?: OptionList): InputJax;
+}
+
+/*****************************************************************/
+/*
+ *  The abstract InputJax class
+ */
+
+export abstract class AbstractInputJax implements InputJax {
+
+    public static NAME: string = 'generic';
+    public static OPTIONS: OptionList = {};
+
+    public options: OptionList;
+    public preFilters: FunctionList;
+    public postFilters: FunctionList;
+
+    /*
+     * @param{OptionList} options  The options to applyt to this input jax
+     *
+     * @constructor
+     */
+    constructor(options: OptionList = {}) {
+        let CLASS = this.constructor as InputJaxClass;
+        this.options = userOptions(defaultOptions({}, CLASS.OPTIONS), options);
+        this.preFilters = new FunctionList();
+        this.postFilters = new FunctionList();
+    }
+
+    /*
+     * @return{string}  The name of this input jax class
+     */
+    public get name() {
+        return (this.constructor as InputJaxClass).NAME;
```

Comment:
It *is* the class member `NAME`, but it is the one for the class of whatever subclass of `AbstractInputJax` is the instance that is calling this method.  That is, this gets the proper name from the class of whatever the instance object is an instance of.